mock/tests/plugins/test_rpmautospec.py

```diff
@@ -0,0 +1,202 @@
+
+from .base import import_plugin
+
+UNSET = object()
+
+
+@pytest.fixture
+def plugin_module():
+    return import_plugin("rpmautospec")
+
+
+def test_init(plugin_module):
+    plugins = object()
+    conf = object()
+    buildroot = object()
+
+    with mock.patch.object(plugin_module, "RpmautospecPlugin") as RpmautospecPlugin:
+        plugin_module.init(plugins, conf, buildroot)
+
+    RpmautospecPlugin.assert_called_once_with(plugins, conf, buildroot)
+
+
+class TestRpmautospecPlugin:
+    DEFAULT_OPTS = {
+        "requires": ["rpmautospec"],
+        "cmd_base": ["rpmautospec", "process-distgit"],
+    }
+
+    def create_plugin(self, plugin_module, plugins=UNSET, conf=UNSET, buildroot=UNSET):
+        if plugins is UNSET:
+            plugins = mock.Mock()
+        if conf is UNSET:
+            conf = deepcopy(self.DEFAULT_OPTS)
+        if buildroot is UNSET:
+            buildroot = mock.Mock()
+
+        return plugin_module.RpmautospecPlugin(plugins, conf, buildroot)
+
+    @pytest.mark.parametrize(
+        "with_cmd_base", (True, False), ids=("with-cmd_base", "without-cmd_base")
+    )
+    def test___init__(self, with_cmd_base, plugin_module):
+        plugins = mock.Mock()
+
+        conf = deepcopy(self.DEFAULT_OPTS)
+        if with_cmd_base:
+            expectation = nullcontext()
+        else:
+            expectation = pytest.raises(ConfigError)
+            del conf["cmd_base"]
+
+        buildroot = mock.Mock()
+
+        with mock.patch.object(plugin_module, "getLog") as getLog, expectation as exc_info:
+            logger = getLog.return_value
+            plugin = self.create_plugin(
+                plugin_module, plugins=plugins, conf=conf, buildroot=buildroot
+            )
+
+        if with_cmd_base:
+            assert plugin.buildroot is buildroot
+            assert plugin.config is buildroot.config
+            assert plugin.opts is conf
+            assert plugin.log is logger
+            plugins.add_hook.assert_called_once_with("pre_srpm_build", plugin.attempt_process_distgit)
+            logger.info.assert_called_once_with("rpmautospec: initialized")
+        else:
+            assert "rpmautospec_opts.cmd_base" in str(exc_info.value)
+
+    @pytest.mark.parametrize(
+        "with_exception", (False, True), ids=("without-exception", "with-exception")
+    )
+    def test_as_root_user(self, with_exception, plugin_module):
+        plugin = self.create_plugin(plugin_module)
+
+        if with_exception:
+            expectation = pytest.raises(Exception)
+        else:
+            expectation = nullcontext()
+
+        with plugin.as_root_user(), expectation:
+            plugin.buildroot.uid_manager.becomeUser.assert_called_once_with(0, 0)
+            plugin.buildroot.uid_manager.restorePrivs.assert_not_called()
+            if with_exception:
+                raise Exception("Something happened!")
+
+        plugin.buildroot.uid_manager.restorePrivs.assert_called_once_with()
```

Comment:
If we drop this function/context-manager (see above) we probably don't need this test.

If we keep it, I'd like us to actually test within this code whether `os.geteuid()` returns 0 (root),.

Reply:
I’ll drop it, but regardless…

It’s a unit test, i.e. whether or not `becomeUser()` does its job is out of scope – this method should have its own unit test, which would open its own can of worms 😉: we would have to grant the test suite privs to become root in the first place, or else it will fail (real mock uses consolehelper but I hesitate doing something like this for unit tests).